distributed/client.py

```diff
@@ -603,14 +616,47 @@ def set_error(self, exception, traceback):
             The exception
         traceback: Exception
             The traceback
+        erred_on: set
+            Workers where it errored
+        key: Hashable
+            Task key
+        run_spec: dict[str, Any]
+            dict with args, kwargs, function
         """
         _, exception, traceback = clean_exception(exception, traceback)
 
+        self.key = key
+        self.erred_on = erred_on
         self.status = "error"
         self.exception = exception
         self.traceback = traceback
+        if run_spec:
+            func_args = run_spec.get("args")
+            func_kwargs = run_spec.get("kwargs")
+            self.func_args = pickle.loads(func_args) if func_args else ()
```

Comment:
@crusaderky `pickle.loads` is probably not the right way to do this. What's the right way?

Reply:
It's complicated. See
https://github.com/dask/distributed/blob/a7f7764ccf68a5efd10b6b3ee7227af49d7000a5/distributed/worker.py#L2206-L2222
and
https://github.com/dask/distributed/blob/a7f7764ccf68a5efd10b6b3ee7227af49d7000a5/distributed/worker.py#L2930-L2947

Please call `distributed.worker._deserialize` from here.

In rare use cases, run_spec may be `None` - for example if scattered data fails to deserialize. Please deal with the use case, but don't lose time writing a unit test for it - there are bugs related to it.